### DISCUSSION TITLE
Using assertions in NCC library functions

### DISCUSSION THREAD

**Discussion by thass0** [maximecb/uvm#31] [General]
Some C library functions don't allow their pointer arguments to be `NULL` (e.g. `strtok`s delimiter argument). In case of `glibc`, they respond to `NULL` by crashing with a segfault. For debugging purposes, and to make it easier to reason about segfaults, it would be beneficial if they called `assert(x != NULL)` instead.
The downside of adding assertions to frequently called functions might be that the performance is slightly worse. At the time of writing this, there are a few assertions found in `/include/uvm` already.

In general, should assertions be added to library functions that have certain constraints like requiring pointers to not be `NULL`?

**Comment by maximecb** [comment]
Yes I think we should add assertions. I don't expect string functions to be super performance-sensitive, and in theory we can eliminate all the overhead of all assertions by compiling code in release mode.

**Reply by thass0** [reply]
Yes, I agree.
